### ISSUE TITLE
BUG???

### ISSUE THREAD

**Issue by Reatris** [bdsaglam/torch-scae#10]
in the file  **object_decoder.py** line 338
`
posterior_mixing_prob = F.softmax(posterior_mixing_logits_per_point, 1)
`

but in the file  **stacked_capsule_auto_encoder.py** line 210
`
mass_explained_by_capsule = res.posterior_mixing_prob.sum(-1)
`

Sum (dim = - 1) this operation makes all values after softmax (dim = 1) become 1.
So is this value useful ?, a tensor with all value is 1.

**Comment by bdsaglam** [comment]
Have you checked the actual values? I am not seeing all ones. ```posterior_mixing_logits_per_point``` tensor has three axes; batch, object capsule, part capsule. The softmax is applied on the object axis. Then, it is summed over on the part axis to compute explainability rate per object capsule.

**Comment by Reatris** [comment]
Thank you for your reply. Although not all values are 1, they are very close to 1. The most important thing is that they all have the same value, which is what I observed. Then I would like to ask a question, if the template size is not 11x11, but is the same as H, W of the input image, what is the result?

**Comment by Reatris** [comment]
The reason why I set up a large template is that the API of my framework can only use the same template as the original image in length and width. I guess this makes my training have problems.

**Comment by bdsaglam** [comment]
> Thank you for your reply. Although not all values are 1, they are very close to 1. The most important thing is that they all have the same value, which is what I observed. Then I would like to ask a question, if the template size is not 11x11, but is the same as H, W of the input image, what is the result?

That's interesting, it does not make sense having all values close to 1 since softmax and sum is applied over different axes. 
The original paper has 11x11 templates and they are affine transformed to image size with a learned transformation matrix. Having large templates may make training slower because of increased number of parameters. By the way, there is already a bug in the implementation so the problem might not be template size.

**Comment by Reatris** [comment]
Thank you very much. I look forward to your update.

**Comment by bdsaglam** [comment]
Fixed in #16. The main bug was that part templates were warped with the pose instead of its inverse, which blew every template up to at least the size of the canvas. With that and a few smaller mismatches fixed, 300k steps on 40x40 MNIST give 98.6–98.7% k-means test accuracy (paper: 98.7%). See the README for results and `torch_scae_experiments/mnist/evaluate.py` for the evaluation. Once parts are local, the posterior mass per object capsule is informative: k-means on it gives 98.6% test accuracy. Templates are meant to be smaller than the image (11x11 on a 40x40 canvas).
